lib/api.rb

```diff
@@ -243,59 +323,157 @@ def validation_error(msg)
     }.to_json
   end
 
-  #
-  # ciphers
-  #
-
-  # create a new cipher
-  post "/ciphers" do
+  # Used by the web vault to update the private and public keys if the user doesn't have one.
+  post "/accounts/keys" do
+    content_type :json
+    # Needed by the web vault for EVERY response
+    response['access-control-allow-origin'] = '*'
     d = device_from_bearer
     if !d
       return validation_error("invalid bearer")
     end
 
-    need_params(:type, :name) do |p|
-      return validation_error("#{p} cannot be blank")
+    if !params[:encryptedprivatekey].to_s.match(/^2\..+\|.+/)
+      return validation_error("Invalid key")
     end
 
-    begin
-      Bitwarden::CipherString.parse(params[:name])
-    rescue Bitwarden::InvalidCipherString
-      return validation_error("Invalid name")
+    d.user.private_key = params[:encryptedprivatekey]
+    d.user.public_key = params[:publickey]
+
+    {
+      "Id" => d.user_uuid,
+      "Name" => d.user.name,
+      "Email" => d.user.email,
+      "EmailVerified" => d.user.email_verified,
+      "Premium" => d.user.premium,
+      "MasterPasswordHint" => d.user.password_hint,
+      "Culture" => d.user.culture,
+      "TwoFactorEnabled" => d.user.totp_secret,
+      "Key" => d.user.key,
+      "PrivateKey" => d.user.private_key,
+      "SecurityStamp" => d.user.security_stamp,
+      "Organizations" => "[]",
+      "Object" => "profile",
+   }.to_json
```

Comment:
again: i think it should be possible to adopt the ```user.to_hash``` function and then use it instead of the manual hash.